bpo-32972: Add unittest.AsyncioTestCase

dave-shawley opened the pull request:
This PR adds co-routine support to unittest by adding a new sub-class of `unittest.TestCase` named `unittest.AsyncioTestCase`.  It extends `unittest.TestCase` by adding the following functionality:

- a new event loop is created and terminated with each test method & the *same* event loop is used throughout the entire test method execution from setup through cleanup
- asynchronous test setup and teardown are supported by new methods - `asyncSetUp` and `asyncTearDown`
- test methods that are decorated with the `async` keyword are run on the event loop.  Undecorated methods are run as normal methods.
- cleanup hooks that are decorated with the `async` keyword are run on the event loop.  Undecorated methods are run as normal methods.

I made the following conscious decisions during the implementation that _may_ be somewhat controversial so I want to mention them here (in no particular order) along with the rationale.

1. **there is no direct support for other event loop implementation**: other event loop implementations can be plugged in using [a custom EventLoopPolicy implementation](https://docs.python.org/3/library/asyncio-policy.html).  It can be inserted by overriding `setUpClass`.  I believe that other loop implementations can either plug in using the event loop policy or provide their own asynchronous test case implementation (e.g., [tornado.testing](http://www.tornadoweb.org/en/stable/testing.html)).
2. **asyncio.iscoroutine is used to detect async methods**: the two places that I need to detect whether a callable is a co-routine or not are: (a) test methods and (b) cleanup hooks added with [addCleanup](https://docs.python.org/3/library/unittest.html#unittest.TestCase.addCleanup).  Both are explicit (opt-in) actions so they are completely controlled by the test writer.
3. **outstanding asyncio Tasks are not terminated**: I'm still on the fence about whether the test should fail if there are outstanding tasks or if it should gather them.
4. **async test case is a separate class**: I decided to use a sub-class so that it is immediately obvious that a test methods are intended to run on the event loop.  This also insulates `unittest.TestCase` from unexpected breakages and maintains it's current behavior.

**Open Questions**
1. What should happen when `len(asyncio.all_tasks())` is non-zero after cleanups are run?
2. Should `unittest.AsyncioTestCase` include a global test run timeout?

I still have to amend commits to update documentation but I wanted to have the discussion about the implementation choices before I started writing up examples and updating the existing documentation suite.

<!-- issue-number: [bpo-32972](https://bugs.python.org/issue32972) -->
https://bugs.python.org/issue32972
<!-- /issue-number -->

dave-shawley committed:
unittest.TestRunner: introduce _runTest method.

This is the core of running a test.  It is also the portion of run()
that requires specialization for asyncio support.

dave-shawley committed:
unittest: Refactor TestCase.debug.

This method is essentially ``_runTest`` with a Outcome object that does
nothing.  This commit inserts a new class that is a no-op Outcome so
that the method can be re-used.

dave-shawley committed:
Add unittest.AsyncioTestCase.

dave-shawley committed:
unittest: Introduce asyncSetUp/TearDown.

See discussion under https://bugs.python.org/issue32972

dave-shawley committed:
unittest: Introduce TestCase._terminateTest.

This makes AsyncioTestCase.doCleanups() safe to call at any time.  The
previous version terminated the event loop unnecessarily.

dave-shawley committed:
unittest: Shutdown async generators.

dave-shawley committed:
unittest: Enable event loop debug flag.

matrixise commented on Lib/unittest/case.py:
```diff
@@ -1439,3 +1454,150 @@ def shortDescription(self):
+    is stopped and closed.
+
+    When subclassing AsyncioTestCase, the event loop is available
+    via the ``event_loop`` property.  Tests can assume that
```
about ``event_loop`` I am not sure, usually we found ``loop`` in the documentation of asyncio

matrixise commented on Lib/unittest/case.py:
```diff
@@ -1439,3 +1454,150 @@ def shortDescription(self):
+        self.__event_loop = None
+
+    def doCleanups(self):
+        """
```
The docstring should start on the first line, https://www.python.org/dev/peps/pep-0008/#documentation-strings

matrixise commented on Lib/unittest/case.py:
```diff
@@ -81,6 +82,16 @@ def testPartExecutor(self, test_case, isTest=False):
             self.success = self.success and old_success
 
 
+class _IgnoredOutcome(object):
```
In Python3, we prefer to use the new style
https://docs.python.org/3/reference/compound_stmts.html#class-definitions
```python
# new style
class _IgnoredOutcome:
    pass
```
is equivalent to 
```python
# old style
class _IgnoredOutcome(object):
    pass
```

so in this case, use the new style and not the old one.

matrixise commented on Lib/unittest/test/test_asyncio_test_case.py:
```diff
@@ -0,0 +1,281 @@
+import unittest
+
+
+class Test(object):
```
See my previous comment about the construction of a class
```python
class Test:
    pass
```
is better

matrixise commented on Lib/unittest/case.py:
```diff
@@ -1439,3 +1454,150 @@ def shortDescription(self):
+        self.tearDown()
+
+    @property
+    def event_loop(self):
```
in the documentation, we prefer to use the `loop` variable for the reference to the event loop.

Examples:
```:meth:`loop.call_soon` ``` or ```:meth:`loop.set_debug` ``` etc...

maybe you should do the same.

matrixise commented on Lib/unittest/test/test_asyncio_test_case.py:
```diff
@@ -0,0 +1,281 @@
+        raise AssertionError(f'test {test_class} unexpectedly failed')
+
+
+class EventLoopTracking(object):
```
just `class EventLoopTracking:`

matrixise reviewed (changes requested):
Thank for your contribution but could you add a NEWS entry with the blurb tool.

`pip install blurb`

dave-shawley committed:
bpo-32972: Switch to "newer style" classes.

See https://github.com/python/cpython/pull/10296

dave-shawley committed:
bpo-32972: Docstring cleanup in unittest.case.

See https://github.com/python/cpython/pull/10296.

dave-shawley committed:
bpo-32972: Rename AsyncioTestCase.event_loop.

`loop` is consistent with the Standard Library spelling of the current
event loop.

See https://github.com/python/cpython/pull/10296

dave-shawley committed:
bpo-32972: Add NEWS blurb.

See https://github.com/python/cpython/pull/10296

dave-shawley committed:
Merge branch 'master' into bpo-32971-async-test-runner

PetterS commented on Lib/unittest/case.py:
```diff
@@ -1483,3 +1486,149 @@ def shortDescription(self):
+        if self.__loop is None:
+            self.__loop = asyncio.new_event_loop()
+            self.__loop.set_debug(True)
+            asyncio.set_event_loop(self.__loop)
```
Does the test case have to modify the global state like this? It is calling `self.loop.run_until_complete` when running things.

dave-shawley replied:
I think that it does.  This is the only way to ensure that the code under test gets the correct loop.
